### DISCUSSION TITLE
[Solution] Authorization system that hides spec url's for unauthorized users

### DISCUSSION THREAD

**Discussion by WilliamStam** [fastapi/fastapi#10692] [Show and tell]
this has been a huge bone of contention for me. its been 3 weeks and 3 frameworks later for me just for this 1 issue (according to my boss its been a day k?)....

this isnt ment to be "production ready" yet. its just a "FFS FINALLY GOT IT" moment. so results may vary

lets break down the requirements:

- we need to attach what permissions the user needs to have to the route when we setup the route. (dev time)
- need to check the current user and allow / disallow them if the current user doesnt have that permission (route run
time)
- we need to hide routes that the user doesnt have permission on (spec gen time)

fastapi doesnt let you add anything to the route object (closest you get is openapi_extras) but that didnt work for me. so long story short.. why not use the Authorization "guard" for this?

```python

router = APIRouter(
    tags=["system"],
    prefix="/system",
    dependencies=[Authorization("system")]
)


@router.get("/profile", dependencies=[Authorization("profile")])
async def get_profile(
    user: CurrentUser,
) -> model.ProfileModel:
    
    return user

```

the user will need a string list for permissions to be able to visit these endpoints ie:

```json
{
  "permissions": [
    "system",
    "profile"
  ]
}
```

lets take a look at the Authorization dependency. 

```python

class AuthorizationDependency:

    def __init__(self, permissions: typing.Union[str, typing.List[str], None] = None):
        self.permissions = permissions

    def get_permissions(self):
        permissions = self.permissions
        if self.permissions is None:
            permissions = []

        if isinstance(permissions, Permission):
            permissions = [permissions]
        elif isinstance(permissions, str):
            permissions = [permissions]
        else:
            permissions = list(permissions)

        return permissions

    async def __call__(
        self,
        request: Request,
        user = Depends(get_current_user)
    ) -> ProfileModel:

        if user.id is None:
            logger.debug("User not logged in")
            raise HTTPException(status_code=401)

        missing_permissions = []
        for permission in self.get_permissions():
            if str(permission) not in user.permissions:
                missing_permissions.append(str(permission))
            
            
        if missing_permissions:
            logger.debug(f"User missing permissions '{missing_permissions}'")
            raise HTTPForbiddenException(status_code=403, permissions=missing_permissions)

        return user


def Authorization(
    permissions: typing.Union[str, typing.List[str], None] = None
) -> Depends :
    return Depends(AuthorizationDependency(permissions))

```

any time you have a dependencies=[Authorization] the system needs to make sure the user is logged in. so a route of `@router.get("/profile", dependencies=[Authorization()])` will require the user to be logged in before they can view the endpoint. you can pass a string or a list of strings to the Authorization dependency for it to check against.

--------
bonus bit: you can take the Authorization dependency to new heights as well. just include "fetching the token" to it and any page that had `Authorization` dependency gets the padlock in swagger

```python
def Authorization(
    permissions: typing.Union[Permission, typing.List[Permission], None] = None, 
    token=Depends(get_token)
) -> Depends :
   ...
```

you arent "using" the token for anything here. its just triggering the padlock part. 

anywhere where i have auth or get the current user it has the padlock because of this:
```python

QUERY_API_KEY = APIKeyQuery(name='token', auto_error=False, description="Pass the token via a query sting item ?token=xxx")
BEARER = HTTPBearer(auto_error=False, description="Use a bearer token for authorization")
HEADER_API_KEY = APIKeyHeader(name='X-API-Key', auto_error=False, description="Add a header [X-API-Key] with the token")

logger = log.getLogger(__name__)

async def get_token(
    query_api_key: str = Security(QUERY_API_KEY),
    bearer_key: HTTPAuthorizationCredentials = Security(BEARER),
    header_api_key: str = Security(HEADER_API_KEY),
):
    if bearer_key:
        bearer_key = bearer_key.credentials
    
    token = next(
        (arg for arg in [query_api_key, bearer_key, header_api_key] if arg is not None),
        None
    )
    
    return token


async def get_current_user(token=Depends(get_token)) -> ProfileModel:
    user = ProfileModel()
    if token:
        user = await find_by_token(token=token)
    
    logger.debug(f"current user: {user}")
    return user


CurrentUser = typing.Annotated[ProfileModel, Depends(get_current_user, use_cache=False)]
```

-------

right so this gets us points 1 and 2. 

we are about 16% of the way there...

the automatic rendering of the openapi spec wont care about a thing we're adding in. so we need to replace it. its not worthy of our awesome any more. 

so first we need to disable the automatic openapi.json generating endpoint

```python

app = FastAPI(
    # boring stuff
    docs_url=None,
    openapi_url=None,
    # boring stuff
)
```

we will need a new route for our docs so basically create something like the following but im not going to walk through it

```python
@router.get('/', response_class=HTMLResponse, include_in_schema=False)
@router.get("/rapidoc", response_class=HTMLResponse, include_in_schema=False)
async def rapidoc(
    request: Request,
    user: CurrentUser
):
    app = request.app
    templates = Jinja2Templates(directory="templates")
    user = user
    
    return templates.TemplateResponse("rapidoc.jinja", {
        "request": request,
        "app": app,
        "user": user
    })
```

right so now onto the "magic"

thought process behind the openapi.json route would be to loop through all the routes, check if the current user (dependency on this route vs permissions set on the route) and then set include_in_schema=False. but this had some interesting results. for 1.. we dont have the permissions on the route? but thanks to the trusty debugger... the route DOES have dependencies on it. and its a list of all dependencies. and since we used an object for authorization we can find all dependencies that has our permission object and then look into it to get the permissions set. SCORE. 

the other issue i faced was with setting the schema hide thing. since the schema is generated on startup and not changed after that.. we need to reset it.
chrisK824 in https://github.com/tiangolo/fastapi/discussions/10264#discussioncomment-7126645 answered that nicely. but hold up. it didnt work so well. it would still cache the schema and laugh at your face. (short story.. adding include_in_schema=False to the route doesnt "show" it again on future visits once the user logged in / permission changed)

so what if we take a page out of the `app.openapi = custom_openapi` book..

and behold

```python


from fastapi.openapi.utils import get_openapi


@router.get('/openapi.json', include_in_schema=False)
async def openapi(request: Request, user: CurrentUser):
    app = request.app
    routes = []
    
    for route in request.app.routes:
        include_in_schema = True
        if hasattr(route, "dependencies") and type(getattr(route, "dependencies")) == list:
            # the route object "might" have a dependencies attribute
            for row in route.dependencies:
                # for each dependency find those that have dependency in it. this part can defs be made better. was hoping for something like if isinstance(Authorization) but at this point its actualy a Depends()
                if hasattr(row, "dependency"):
                    # its hot and i went for shortcuts ok... see previous comment
                    if hasattr(row.dependency, "permissions"):
                        # these are the permissions set on the route
                        permissions = row.dependency.permissions
                        # same as the Auth logic, if the auth dependency is found then a user is required
                        if user.id is None:
                            include_in_schema = False
                        # if the user doesnt have all the permissions then we need to hide it
                        if not user.has_permissions(permissions):
                            include_in_schema = False
        
        # we now create a "new" 'sequence' of routes
        if include_in_schema:
            routes.append(route)
        
    
    # create a new openapi object with out new routes in
    openapi_schema = get_openapi(
        title=app.title,
        version=app.version,
        openapi_version=app.openapi_version,
        description=app.description,
        terms_of_service=app.terms_of_service,
        contact=app.contact,
        license_info=app.license_info,
        routes=routes, # our new routes
        tags=app.openapi_tags,
        servers=app.servers
    
    )
    
    return JSONResponse(openapi_schema)
```

and voila. 

i honestly cant believe that there is literally NOTHING available on this type of thing anywhere that ive looked. everyone just seems to be "all docs or no docs" all the time

**Comment by acidjunk** [comment]
Great stuff: I wonder is this still openapi compliant? I am used to be able to create/autogenerate a client: that normally happens without authentication/authorization.

**Reply by WilliamStam** [reply]
It purely hides routes. So according to the front end it's like they were never there. There is 1 big front end "issue" tho. Since you need to authenticate with the API to get the API docs json, the rapidoc swagger etc doesn't re fetch the openapi.json file when you Auth in the interface and send the token along the same way they do for "try it" stuff. So you would probably have to handle setting token outside of the interface like have a login screen, capture token, send that token with the spec-url fetch.

**Reply by acidjunk** [reply]
The tools that generate a client look at the openapi.yaml/json: so they would probably fail.
I like the approach though; was previously thinking to add some sort of an API portfolio in front of FastAPI top achieve what you did.

**Comment by WilliamStam** [comment]
they dont fail. to them its like you simply didnt include those routes to the application in the first place. the only issue is that since the openapi.json endpoint needs to auth the user by default only the "open" routes are visible to swagger / redoc / rapiddoc / elements 

unauthed (no token set in the address bar. the token just gets sent through via templating to the spec-url)
rapidoc - 
![image](https://github.com/tiangolo/fastapi/assets/182800/726412d8-131f-45c9-87a8-9f0a2a6387ff)
elements - 
![image](https://github.com/tiangolo/fastapi/assets/182800/c61482dd-465f-4991-8c54-e6ed4c06b987)
the horrible 1 (swagger)
![image](https://github.com/tiangolo/fastapi/assets/182800/1511cfa0-4d24-4d8c-8a99-a01c031e0cfb)
redoc -
![image](https://github.com/tiangolo/fastapi/assets/182800/f5414151-85ac-4d07-8dca-c1b3c0f0219e)

now im only gonna add ?token=123 to the address bar (ive made my token 123 for testing purposes. usualy its uuid)

rapidoc
![image](https://github.com/tiangolo/fastapi/assets/182800/1cb7dff3-dcb8-43c9-98e5-6029030a4c32)

elements
![image](https://github.com/tiangolo/fastapi/assets/182800/6e2244db-29f2-49d0-a9b4-1c8768864ba3)

yuky 1
![image](https://github.com/tiangolo/fastapi/assets/182800/b1c6c60d-d0b4-4a20-9bdd-b03d5ac9bbce)

redoc 
![image](https://github.com/tiangolo/fastapi/assets/182800/5131867a-4ba7-48fc-8a48-e9987f0af2a3)

-------

```


@router.get("/", response_class=HTMLResponse, include_in_schema=False)
@router.get("/rapidoc", response_class=HTMLResponse, include_in_schema=False)
async def rapidoc(request: Request, user: CurrentUser):
    app = request.app
    templates = Jinja2Templates(directory="templates")
    return templates.TemplateResponse("rapidoc.jinja", {"request": request, "app": app, "user": user})

@router.get("/elements", response_class=HTMLResponse, include_in_schema=False)
async def elements(request: Request, user: CurrentUser):
    app = request.app
    templates = Jinja2Templates(directory="templates")
    return templates.TemplateResponse("elements.jinja", {"request": request, "app": app, "user":user})

@router.get("/swagger", response_class=HTMLResponse, include_in_schema=False)
async def swagger(request: Request, user: CurrentUser):
    app = request.app
    templates = Jinja2Templates(directory="templates")
    return templates.TemplateResponse("swagger.jinja", {"request": request, "app": app, "user": user})

@router.get("/redoc", response_class=HTMLResponse, include_in_schema=False)
async def redoc(request: Request, user: CurrentUser):
    app = request.app
    templates = Jinja2Templates(directory="templates")
    return templates.TemplateResponse("redoc.jinja", {"request": request, "app": app, "user": user})

```

rapidoc
```html
<!doctype html>
<html>
<head>
    <meta charset="utf-8">
    <script type="module" src="/static/rapidoc-min.js"></script>
    <script src="/static/rapipdf-min.js"></script>
    <title>{{ app.title }}</title>
</head>
<body>
<rapi-doc
        id="docs"
        spec-url="/openapi.json?token={{ user.token }}"
        theme="light"
        heading-text="{{ app.title }}"
        primary-color="#ff8300"
        nav-bg-color="#212322"
        nav-accent-color="#ff8300"
        nav-accent-text-color="#ff8300"
        persist-auth="true"
        show-method-in-nav-bar="as-colored-block"

        layout="column"
        render-style="focused"
        nav-item-spacing="compact"
        show-header="false"
        allow-server-selection="false"
        allow-search="false"
        allow-advanced-search="false"
        use-path-in-nav-bar="false"

        sort-tags="true"

>
    <span slot="logo"></span>
    <div slot="nav-logo">


    </div>
    <div slot="overview">
        
    </div>


</rapi-doc>

</body>
</html>

```

swagger
```html
<!DOCTYPE html>
<html>
<head>
    <title>{{ app.title }}</title>
    <meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link href="https://cdn.jsdelivr.net/npm/swagger-ui-dist@5.1.3/swagger-ui.css" rel="stylesheet">
    <script src="https://cdn.jsdelivr.net/npm/swagger-ui-dist@5.1.3/swagger-ui-bundle.js" crossorigin></script>
    <script src="https://cdn.jsdelivr.net/npm/swagger-ui-dist@5.1.3/swagger-ui-standalone-preset.js" crossorigin></script>
</head>
<body>
<div id='swagger-container'/>
<script type="text/javascript">
    const ui = SwaggerUIBundle({
                url: "/openapi.json?token={{ user.token }}",
                dom_id: '#swagger-container',
                deepLinking: true,
                showExtensions: true,
                showCommonExtensions: true,
                presets: [
                    SwaggerUIBundle.presets.apis,
                    SwaggerUIBundle.SwaggerUIStandalonePreset
                ],
            })

</script>
</body>
</html>

```

**Reply by WilliamStam** [reply]
the only difference between the before and after auth is the inclusion of ?token=123 in the address bar. which finds a valid user that has permissions for all those endpoints. without a valid user it hides them again and only "public" endpoints are visible again 
![image](https://github.com/tiangolo/fastapi/assets/182800/a8aa8449-9760-4d62-b828-fced04700357)

-------

the issue of how to take the token specified in the "front end" (and change the call to the /openapi.json including the set token) is a totaly other debate and not something thats under the scope of fastapi. that would be something like on capture auth method, use the same mechanism for "try it" to the spec url thing and reload the spec url. 

![image](https://github.com/tiangolo/fastapi/assets/182800/4309571c-6d2b-41e6-8d9c-7747cc1852a6)

and include it and reload

![image](https://github.com/tiangolo/fastapi/assets/182800/e4b766db-e39b-43cb-b306-0603c3cdb323)
